drivers/rtc: Add STM32U3 RTC support and enable RTC on nucleo_u385rg_q board

AntoinePST opened the pull request:
This pull request adds comprehensive support for the RTC peripheral on STM32U3 series devices.

Changes include:
- Add RTC node in the STM32U3 device tree source (dts) to enable RTC hardware description.
- Extension of the STM32 RTC driver to handle STM32U3-specific behaviors, including alarm interrupt routing and clock enabling.
- Enabling RTC support on the nucleo_u385rg_q board by updating its device tree and board configuration files.
- Addition of RTC API test configuration and overlay for the nucleo_u385rg_q board to validate RTC functionality.

These changes collectively enable RTC functionality on STM32U3 platforms and the nucleo_u385rg_q development board, facilitating accurate timekeeping and alarm features.

Signed-off-by: Antoine Pradoux <antoine.pradoux@st.com>

AntoinePST committed:
dts: arm: st: Add RTC node for STM32U3 series

- Add the rtc node for the stm32u3
- This enables RTC support for STM32U3 platforms once enabled

Signed-off-by: Antoine Pradoux <antoine.pradoux@st.com>

AntoinePST committed:
boards: st: Enable RTC support on nucleo_u385rg_q board

- Activate the RTC node in the device tree by enabling its clocks and
setting its status to okay
- Add 'rtc' to the supported features in the board YAML configuration

Signed-off-by: Antoine Pradoux <antoine.pradoux@st.com>

AntoinePST committed:
drivers: rtc: stm32: Add STM32U3 support for RTC alarm, clock & counter

Extend RTC driver to handle STM32U3 series specifics:
- Include STM32U3X in condition where RTC Alarm event is not routed to EXTI
- Use LL_RCC_RTC_ClockEnable call for STM32U3 instead of LL_RCC_EnableRTC
- Adapt counter_stm32_rtc driver for STM32U3 clock and EXTI handling

Signed-off-by: Antoine Pradoux <antoine.pradoux@st.com>

AntoinePST committed:
tests: rtc_api: Add RTC test configuration for nucleo_u385rg_q board

- Add a new test configuration file enabling RTC calibration and alarm
features for the nucleo_u385rg_q board
- Add the nucleo_u385rg_q device tree overlay defining the RTC alias

Signed-off-by: Antoine Pradoux <antoine.pradoux@st.com>

mathieuchopstm commented on drivers/rtc/rtc_stm32.c:
```diff
@@ -506,8 +506,10 @@ static int rtc_stm32_init(const struct device *dev)
  * On STM32WB0 series, LL_RCC_EnableRTC is not provided by STM32CubeWB0,
  * but RTC IP clock has already been turned on - skip the call as well.
+ * On STM32U3 series, LL_RCC_EnableRTC is not provided by STM32CubeU3,
+ * but RTC IP clock has already been turned on - skip the call as well.
```
Outside PR scope: I worded this extremely poorly when introducing WB0 series.
I propose to change as follows:
```c
/*
 * On certain series, there is no bit in BCDR register to enable RTC;
 * the RTC is enabled by the same RCC bit that enables the bus interface
 * (RTCEN). On such series, the LL_RCC_EnableRTC function is usually not
 * provided by the STM32Cube package, but it's fine to skip calling it
 * since the RTC has already been enabled by clock_control_on() above.
 */
```

djiatsaf-st commented on tests/drivers/rtc/rtc_api/boards/nucleo_u385rg_q.overlay:
```diff
@@ -0,0 +1,11 @@
+/*
+ * Copyright (c) 2024 STMicroelectronics
```
```suggestion
 * Copyright (c) 2025 STMicroelectronics
```

AntoinePST commented:
This update includes the following corrections:
- Removed the unnecessary style fix commit by integrating those changes properly into the original commits.
- Corrected the conditional macros in drivers/rtc/rtc_stm32.c to properly include the STM32U3X series in the RTC Alarm interrupt enable/disable conditions.
- Updated the RTC clock enable function for the STM32U3 series by replacing the call to LL_RCC_EnableRTC() with LL_RCC_RTC_ClockEnable() to comply with the Cube-specific behavior for this series.
- Added detailed comments explaining this hardware-specific quirk in the code.
- Change the year with "2026" in the tests/drivers/rtc/rtc_api/boards/nucleo_u385rg_q.overlay file.
- Maintained consistent code style (indentation, line breaks).
- Other commits (device tree, board configuration, tests) remain unchanged.
- All these modifications have been cleanly integrated into the original commits to keep a clear and readable history.

Thank you for your reviews and support!

Best regards,
Antoine

erwango commented on drivers/rtc/rtc_stm32.c:
```diff
@@ -502,12 +504,17 @@ static int rtc_stm32_init(const struct device *dev)
+ * by the STM32Cube package, but it's fine to skip calling it since
+ * the RTC is already accessible thanks to clock_control_on() above.
+ * On STM32U3 series, RTC enable is done via RCC_BDCR and RCC_APB1ENR1 bits,
+ * and the Cube function is named LL_RCC_RTC_ClockEnable().
```
Actually `LL_RCC_RTC_ClockEnable` and `LL_RCC_EnableRTC` are strictly equivalent (`SET_BIT(RCC->BDCR, RCC_BDCR_RTCEN);`)
My view is that this is an LL API issue that should be fixed in HAL U3 package.

Meanwhile, I'll drop the comment and would go with a single :

```
#if defined(CONFIG_SOC_SERIES_STM32U3X)
	LL_RCC_RTC_ClockEnable();
#else
	LL_RCC_EnableRTC();
#endif
```

Dont care about the `z_stm32_hsem_lock`, they'll be compiled out on U3

etienne-lms commented on boards/st/nucleo_u385rg_q/nucleo_u385rg_q.dts:
```diff
@@ -208,6 +208,11 @@
 };
 
+&rtc {
+	clocks = <&rcc STM32_CLOCK(APB1, 30) &rcc STM32_SRC_LSE RTC_SEL(1)>;
```
Prefer:
```c
		clocks = <&rcc STM32_CLOCK(APB1, 30)>,
		 	 <&rcc STM32_SRC_LSE RTC_SEL(1)>;

etienne-lms commented on drivers/counter/counter_stm32_rtc.c:
```diff
@@ -795,7 +812,8 @@ static int rtc_stm32_init(const struct device *dev)
-#elif defined(CONFIG_SOC_SERIES_STM32U5X) || defined(CONFIG_SOC_SERIES_STM32WBAX)
+#elif defined(CONFIG_SOC_SERIES_STM32U3X) || defined(CONFIG_SOC_SERIES_STM32U5X) || \
+	  defined(CONFIG_SOC_SERIES_STM32WBAX)
 	/* in STM32U5 family RTC is not connected to EXTI */
```
Here also?
```patch
-	/* in STM32U5 family RTC is not connected to EXTI */
+	/* RTC is not connected to EXTI for these SoC series */

AntoinePST replied:
you're right I forgot it sorry

etienne-lms reviewed (commented):
To prevent build failure on commit *"boards: st: Enable RTC support on nucleo_u385rg_q board"* (`LL_RCC_EnableRTC()` not implemented for U3), could you move commit *"drivers: rtc: stm32: Add STM32U3 support for RTC alarm, clock & counter"* before in your series?

Otherwise, LGTM.

bjarki-andreasen commented:
ping @mathieuchopstm

mathieuchopstm reviewed (approved):
Sorry for missing the pings :sweat: - LGTM
